bundles/org.openhab.binding.plugwiseha/src/main/java/org/openhab/binding/plugwiseha/internal/api/model/PlugwiseHAController.java

```diff
@@ -0,0 +1,427 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.plugwiseha.internal.api.model;
+
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerConfigurationException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.stream.StreamSource;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.openhab.binding.plugwiseha.internal.api.exception.PlugwiseHAException;
+import org.openhab.binding.plugwiseha.internal.api.model.dto.*;
+import org.openhab.binding.plugwiseha.internal.api.xml.PlugwiseHAXStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link PlugwiseHAController} class provides the interface to the Plugwise
+ * Home Automation API and stores/caches the object model for use by the various
+ * ThingHandlers of this binding.
+ * 
+ * @author B. van Wetten - Initial contribution
+ */
+@NonNullByDefault
+public class PlugwiseHAController {
+
+    // Private member variables/constants
+
+    private static final int MAX_AGE_MINUTES_REFRESH = 10;
+    private static final int MAX_AGE_MINUTES_FULL_REFRESH = 30;
+    private static final DateTimeFormatter FORMAT = DateTimeFormatter.RFC_1123_DATE_TIME; // default Date format that
+                                                                                          // will be used in conversion
+
+    private final Logger logger = LoggerFactory.getLogger(PlugwiseHAController.class);
+
+    private final HttpClient httpClient;
+    private final PlugwiseHAXStream xStream;
+    private final Transformer domainObjectsTransformer;
+
+    private final String host;
+    private final int port;
+    private final String username;
+    private final String smileId;
+
+    private @Nullable ZonedDateTime gatewayUpdateDateTime;
+    private @Nullable ZonedDateTime gatewayFullUpdateDateTime;
+    private @Nullable DomainObjects domainObjects;
+
+    public PlugwiseHAController(HttpClient httpClient, String host, int port, String username, String smileId)
+            throws PlugwiseHAException {
+        this.httpClient = httpClient;
+        this.host = host;
+        this.port = port;
+        this.username = username;
+        this.smileId = smileId;
+
+        this.xStream = new PlugwiseHAXStream();
+        this.domainObjectsTransformer = PlugwiseHAController
+                .setXSLT(new StreamSource(getClass().getClassLoader().getResourceAsStream("domain_objects.xslt")));
```

Comment:
Did you test whether this works when your binding is loaded by Karaf? I.e. when you drop your jar into the addons folder.

Reply:
Tested and worked flawless